src/comm.ts

```diff
@@ -0,0 +1,404 @@
+    return new Promise((resolve, reject) => {
+      future.onReply = (msg) => {
+        resolve(msg.content);
+      }
+    });
+  }
+
+  private _handleKernelMsg(kernel: IKernel, msg: IKernelMessage): void {
+    if (!validateCommMessage(msg)) {
+      console.error('Invalid comm message');
+      return;
+    }
+    switch(msg.header.msg_type) {
+       case 'comm_open':
+         this._handleOpen(msg);
+         break;
+       case 'comm_msg':
+         this._handleMsg(msg);
+         break;
+       case 'comm_close':
+         this._handleClose(msg);
+         break;
+    }
+  }
+
+  /**
+   * Handle 'comm_open' kernel message.
+   */  
+private _handleOpen(msg: IKernelMessage): void {
+  var content = msg.content;
+  var promise = loadTarget(
+    content.target_name, content.target_module, this._targets
+  ).then((target: (comm: IComm, data: any) => any) => {
+    var comm = new Comm(
+      content.target_name, 
+      content.comm_id, 
+      this._kernel,
+      () => { this._unregisterComm(content.comm_id); }
+    );
+    try {
+      var response = target(comm, content.data);
+    } catch (e) {
+      comm.close();
+      this._unregisterComm(comm.commId);
+      console.error("Exception opening new comm");
+      return Promise.reject(e);
+    }
+    this._commPromises.delete(comm.commId);
+    this._comms.set(comm.commId, comm);
+    return comm;
+  });
+  this._commPromises.set(content.comm_id, promise);
+}
+
+  /**
+   * Handle 'comm_close' kernel message.
+   */  
+  private _handleClose(msg: IKernelMessage): void {
+    var content = msg.content;
+    var promise = this._commPromises.get(content.comm_id);
+    if (!promise) {
+      var comm = this._comms.get(content.comm_id);
+      if (!comm) {
+        console.error('Comm not found for comm id ' + content.comm_id);
+        return;
+      }
+      promise = Promise.resolve(comm);
+    }
+    promise.then((comm) => {
+      this._unregisterComm(comm.commId);
+      try {
+        comm.close(msg.content.data);
+      } catch (e) {
+        console.log("Exception closing comm: ", e, e.stack, msg);
+      }
+    });
+  }
+
+  /**
+   * Handle 'comm_msg' kernel message.
+   */  
+  private _handleMsg(msg: IKernelMessage): void {
+    var content = msg.content;
+    var promise = this._commPromises.get(content.comm_id);
+    if (!promise) {
+      var comm = this._comms.get(content.comm_id);
+      if (!comm) {
+        console.error('Comm not found for comm id ' + content.comm_id);
+        return;
+      } else {
+        var onMsg = comm.onMsg;
+        if (onMsg) onMsg(msg.content.data);
+      }
+    } else {
+      var newPromise = promise.then((comm) => {
+        try {
+          var onMsg = comm.onMsg;
+          if (onMsg) onMsg(msg.content.data);
+        } catch (e) {
+          console.log("Exception handling comm msg: ", e, e.stack, msg);
+        }
+        return comm;
+      });
+      this._commPromises.set(content.comm_id, newPromise);
```

Comment:
there's no need to reset the promise here. You can `.then` a promise multiple times.